downloader/src/main/kotlin/VcsHost.kt

```diff
@@ -44,6 +46,52 @@ enum class VcsHost(
      */
     vararg supportedTypes: VcsType
 ) {
+    AZURE_DEVOPS("dev.azure.com", VcsType.GIT) {
+        private val commitPrefix = "GC"
+
+        override fun getUserOrOrgInternal(projectUrl: URI) = projectUrlToUserOrOrgAndProject(projectUrl)?.first
+
+        override fun getProjectInternal(projectUrl: URI) = projectUrl.path.substringAfterLast("/")
+
+        override fun toVcsInfoInternal(projectUrl: URI): VcsInfo {
+            val url = projectUrl.toString().replace(projectUrl.query.toString(), "").dropLast(1)
+            val revision =
+                projectUrl.getQueryParameterValue("version").firstOrNull()?.withoutPrefix(commitPrefix).orEmpty()
+            val path = projectUrl.getQueryParameterValue("path").firstOrNull()?.withoutPrefix("/").orEmpty()
+
+            return VcsInfo(VcsType.GIT, url, revision, path)
+        }
+
+        override fun toPermalinkInternal(vcsInfo: VcsInfo, startLine: Int, endLine: Int): String {
+            val actualEndLine = if (endLine != -1) endLine else startLine + 1
+
+            val lineQueryParam = "line=$startLine&lineEnd=$actualEndLine&lineColumn=1&lineEndColumn=1"
+            val pathQueryParam = if (vcsInfo.path.isNotEmpty()) "&path=/${vcsInfo.path}" else ""
+            val revisionQueryParam = if (vcsInfo.revision.isNotEmpty()) {
+                "&version=$commitPrefix${vcsInfo.revision}"
+            } else ""
+
+            return "${vcsInfo.url}?$lineQueryParam$pathQueryParam$revisionQueryParam"
+        }
+
+        override fun toArchiveDownloadUrlInternal(userOrOrg: String, project: String, vcsInfo: VcsInfo): String {
+            val team = URI(vcsInfo.url).path.split("/")[2]
```

Comment:
Would it make sense to guard against index-out-of-bounds by using a path iterator, maybe? Similar below.

Reply:
Not sure if this is the nicest way to do this, but I tried.